### PR TITLE
Bugfix/Multiple dataloaders

### PR DESCRIPTION
**Pull request by justusschock**
## What does this PR do?

Fixes an issue where multiple loaders were not parsed correctly (discovered by @carmocca )

## Before submitting
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by Borda — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

File: tests/trainer/test_dataloaders.py [hunk 1]
```diff
@@ -26,6 +26,7 @@
 import tests.helpers.pipelines as tpipes
 from pytorch_lightning import Callback, seed_everything, Trainer
 from pytorch_lightning.callbacks import ModelCheckpoint
+from pytorch_lightning.trainer.states import TrainerState
```
Comment (akihironitta) [inline tests/trainer/test_dataloaders.py:29]: `TrainerState` is unused.
```suggestion
```

File: pytorch_lightning/trainer/supporters.py [hunk 2]
```diff
@@ -294,6 +291,34 @@ def _calc_num_data(datasets: Union[Sequence, Mapping], mode: str) -> Union[int,
...
 
+    def _get_len_recursive(self, data):
+        if isinstance(data, Dataset):
+            return len(data)
+
+        elif isinstance(data, (float, int)):
+            return data
+
+        elif isinstance(data, Mapping):
+            if isinstance(list(data.values())[0], (Mapping, Sequence, Dataset, Iterable)):
```
Comment (tchaton) [inline pytorch_lightning/trainer/supporters.py:302]: Why only index 0 is being checked ?

File: pytorch_lightning/trainer/supporters.py [hunk 3]
```diff
@@ -294,6 +291,34 @@ def _calc_num_data(datasets: Union[Sequence, Mapping], mode: str) -> Union[int,
...
+                return {k: self._get_len_recursive(v) for k, v in data.items()}
+            else:
+                return self._get_len(data)
+        elif isinstance(data, Sequence):
+            data = list(data)
+            if isinstance(data[0], (Mapping, Sequence, Dataset, Iterable)):
```
Comment (tchaton) [inline pytorch_lightning/trainer/supporters.py:308]: same here.

Comment (tchaton) [reply pytorch_lightning/trainer/supporters.py:308]: `any(isinstance(v, (Mapping, Sequence, Dataset, Iterable)) for v in data)` no ?

File: pytorch_lightning/trainer/supporters.py [hunk 4]
```diff
@@ -294,6 +291,34 @@ def _calc_num_data(datasets: Union[Sequence, Mapping], mode: str) -> Union[int,
 
         return length
 
+    def _get_len_recursive(self, data):
```
Comment (edgarriba) [inline pytorch_lightning/trainer/supporters.py:294]:
```suggestion
    def _get_len_recursive(self, data) -> int:
```

File: pytorch_lightning/trainer/supporters.py (hunk 3 above)
Comment (justusschock) [reply pytorch_lightning/trainer/supporters.py:308]: I wanted to avoid iterating over the whole thing twice @tchaton

File: pytorch_lightning/trainer/supporters.py [hunk 5]
```diff
@@ -294,6 +291,34 @@ def _calc_num_data(datasets: Union[Sequence, Mapping], mode: str) -> Union[int,
...
+            data = list(data)
+            if isinstance(data[0], (Mapping, Sequence, Dataset, Iterable)):
+                return [self._get_len_recursive(v) for v in data]
+            else:
+                return self._get_len(data)
+
+        return self._get_len(data)
+
+    @staticmethod
+    def _get_len(dataset):
```
Comment (edgarriba) [inline pytorch_lightning/trainer/supporters.py:316]:
```suggestion
    def _get_len(dataset) -> int:
```

File: pytorch_lightning/trainer/supporters.py [hunk 6]
```diff
@@ -294,6 +291,34 @@ def _calc_num_data(datasets: Union[Sequence, Mapping], mode: str) -> Union[int,
...
+            return len(data)
+
+        elif isinstance(data, (float, int)):
+            return data
+
+        elif isinstance(data, Mapping):
+            if any(isinstance(v, (Mapping, Sequence, Dataset, Iterable)) for v in data.values()):
+                return {k: self._get_len_recursive(v) for k, v in data.items()}
+            else:
+                return self._get_len(data)
```
Comment (carmocca) [inline pytorch_lightning/trainer/supporters.py:305]: no need for this since you have the one at the bottom

```suggestion
```

File: pytorch_lightning/trainer/supporters.py [hunk 7]
```diff
@@ -294,6 +291,34 @@ def _calc_num_data(datasets: Union[Sequence, Mapping], mode: str) -> Union[int,
...
+        elif isinstance(data, Sequence):
+            data = list(data)
+            if any(isinstance(v, (Mapping, Sequence, Dataset, Iterable)) for v in data):
+                return [self._get_len_recursive(v) for v in data]
+            else:
+                return self._get_len(data)
```
Comment (carmocca) [inline pytorch_lightning/trainer/supporters.py:311]: No need for this since you have the one below

```suggestion
```

**Review by carmocca — APPROVED** [review]
